Adapter-binder Design Option

OwenArnold opened the pull request:
My first design option for the Plankton communication layer.

OwenArnold committed:
Adapter-binder design

OwenArnold committed:
Justification and sub-headings

OwenArnold committed:
Ideas for EnvironmentAdapter

OwenArnold committed:
Address Michael H new-user requriement with some input from Michael W.

OwenArnold committed:
Fixed environment updates

OwenArnold committed:
Make Hardware concept clearer

MichaelWedel commented on design/plankton/CommunicationLayerDesign.md:
```diff
@@ -36,13 +36,161 @@ The IBEX team at ISIS, for example, have expressed interest in using Plankton. N
+
+#### The Device-Adapter Types
+
+The system should address the needs of the instrument-data group as a priority. That is to implement robust simulators meeting testable and maintainable internal standards. For this device and adapter separation is vital. In this mode of operation the `Device` represents the state-machine independent of the `Adapter` and the `Adapter` turns the `Device` into the `Hardware-Device` which both behav ...
```
I like the separation between `Device` and `Hardware-Device`. This describes very accurately what I meant yesterday when I talked about "the model of a device" (this would be the `Device` here). I agree in that this separation opens up good ways for testing.

MichaelWedel commented on design/plankton/CommunicationLayerDesign.md:
````diff
@@ -36,13 +36,161 @@ The IBEX team at ISIS, for example, have expressed interest in using Plankton. N
+        
+```
+
+After discussion with Michael Wedel, I am now strongly leaning towards providing a simple single protocol for the `Environment Control`. While `Adapter` flexibility and variation is vital in order to support different device types. the `Environment` is internal to plankton and we should publish and control a single mechanism for this. Michael Wedel has suggested JSON over zeromq.
````
One factor here is that supporting more than one protocol for this takes time that we can't invest into the device simulation framework itself. If the protocol is an issue for a specific application (no zmq support, no JSON support), it's possible to write a "middle man" (just like we did with the EPICS adapter for the live data prototype) for that application, but I don't think this needs to be a part of plankton.

OwenArnold replied:
I think we should do the minimal possible to control a fixed number of parameters `delta-t` etc via a fixed protocol. If we find that there's a specific client needing a specific protocol then we can cross that bridge closer to the time. The only thing I think we should be aware of is portability and easy of obtaining the supporting packages. For example it should be easy to pip install `zeromq`  (I've never used it myself) and simple to write a service through it to control the `Environment`.

MichaelWedel replied:
Yes, it's as simple as `import zmq` and then you setup a connection and send JSON.

MikeHart85 replied:
ZeroMQ + JSON sounds more like a fully fledged API than a control interface. Is this really what we want/need?

Going this route would mean that people would have to write a python script, use (and know how to) `zmq`, and build JSON objects/strings just to, for example, adjust simulation speed. This seems a bit much for the typical "I have a running simulation and want to tweak a few things" scenario.

I think we should consider something more accessible. A telnet-based interface that implements a simple console would be accessible from any Windows or *nix device without further dependencies. Alternatively, we could consider providing a remote python console.

OwenArnold replied:
I think we are converging on the idea of supporting a single-protocol, or at least, not for the time being allowing users to customise the protocol. I think that what we support needs to be portable, easy to install, and very easy to use. Beyond that I'm not too worried.

MikeHart85 replied:
The IBEX team have raised an issue about adding a telnet based "backdoor" to their simulators:

https://github.com/ISISComputingGroup/IBEX/issues/1526

What they're looking for goes a bit beyond what we discussed here (they want to control not just environmental variables like delta t rates, but also override variables inside the simulator itself).

MichaelWedel replied:
If the problem is building JSON then I think we might be better off implementing the ZMQ+JSON anyway and provide a simple client. This is exactly what we did for the live data prototype [here](https://github.com/DMSC-Instrument-Data/live-data-prototype/blob/master/control.py).
Implementing something that a telnet client can talk to requires us to specify a protocol with a reasonable syntax that doesn't limit what we might want to do with this in the future. And no matter what syntax we choose here, somebody will for sure be unsatisfied with it. User input needs to be validated in some fashion as well. Because in the end we also expose the same API this way, it's just more complicated to implement this on the server side.

MichaelWedel replied:
The basic design of this in the live data prototype is quite simple and as I've said before, in my opinion we can just use that. Any object that exposes a runtime user-adjustable parameter inherits from a class called `Controllable`. It needs to implement a function that returns a dict describing what to expose and what that is (property, function, ...). Then there is a server that knows about these `Controllable` objects and can a) answer queries about what it is exposing and b) forward commands to the appropriate `Controllable`. This communication is done in JSON.
I can't advertise this enough, because it is a solution that does exactly what we are discussing here in a quite nice way that doesn't require too much code on the server side and does many things "automatically". Even the "backdoor" thing would not be a big problem with this implementation...

OwenArnold replied:
Thank's @MichaelWedel. I think that we need to have a look at that design here at ISIS. I'll ask Isabella where the telenet request comes from and how important it is to be telnet. I think we should do both of those things today.

OwenArnold replied:
As I said I would do earlier on this morning:
1. I've talked to Isabella via her Issue. telnet is not mandatory, but she does want to be able to set key-value pairs. It turns out they want to be able to automated test their production IOCs (Novel Idea!) with the help of the emulators, for which you actually need to have two concepts (I was trying to keep them separate). (a) The ability to modify, connect, disconnect, and modify the emulator as part of the test scenario. (b) The ability to run the emulator and get back host + port information so that you connect to it. 
2. I've had a look at the `Controllable` work that @MichaelWedel did for the live listener. Here are my thoughts.
   1. I think the `control.py` interface is pretty nice. And actually more convienient than telnet IMHO. I had originally thought that the user would have to construct the JSON by hand, which is one of the reasons I was initially not in favour of this approach.
   2. I like the auto-discovery on the server and the runtime introspection.
   3. Seems like you had to write quite a bit of code for this, so maybe finding a way to reduce the amount of internal code JSON-RPC maybe would be a good idea.
   4. It would be good if we could get some code sharing between the live data prototype and plankton. Approaches seems similar, but there should only be one `Controllable` per device. Not sure how this deviates from the live listener.
   5. It's good that you have some experience in this area. And most of the code looks like it would work for what we want to do for a manual/commandline driven process.

MikeHart85 replied:
@MichaelWedel 

I took a look at the solution in live-data-prototype, and I really like the idea. I think you're right, this is a better solution than a telnet interface. If we ever really need a telnet interface for something, we could build it on top of it.

In addition to the things Owen mentioned, I'd consider how this might be used from a Python script (like a unit test that wants to control the simulation). It would be good to be able to just import a module and use it as a client directly. The `control.py` (or whatever we call the command-line interface) could then use that as well / wrap around it.

That's already almost what's there, it's just that:
- It's a bit awkward that you need two imports / instances (parameter_control_client + general_parameter_controller)
- Some of the controller's methods print to stdout rather than returning information

MikeHart85 replied:
Another thing to consider is how the control server will fit in with our Framework.

The way it is implemented in live-data-prototype, the server has a `run()` with an infinite loop, which is executed in its own thread. We've been looking to avoid threading in order to ensure that the simulation is deterministic. Would it be possible to have the server process requests the same way our adapters do?

Edit: Ah, looks like you've thought about this as well... hadn't seen your last reply.

MichaelWedel replied:
I've been thinking about this a bit. If the simulation and all its aspects are run in one thread and the "command server" in another, does it affect the determinism of the simulation?

On the other hand, it's possible to make everything "process" as well. Maybe we should look at decoupling the processing loops of the control server and the rest, as the control server might need to run on a faster time scale.

OwenArnold replied:
I like the deterministic nature of @MikeHart85 original design and the fact that you don't have to take special measures for race conditions. I would try to preserve that if possible. One option would be to make the "process" tick occur on some integer multiple of the "control" tick.

MikeHart85 replied:
> If the simulation and all its aspects are run in one thread and the "command server" in another, does it affect the determinism of the simulation?

Theoretically, yes... in practice, I can't think of a scenario where it would matter (which doesn't mean there isn't one).

By introducing a separate thread for the command server, we can no longer be sure when exactly (on which cycle) the command takes effect, if the command is issued while the simulation is right on the edge between two cycles. 

However, when running a simulation normally, the determinism aspect isn't that important... after all, there are many non-deterministic factors acting on the simulation from outside (not least of which the user themselves). And when running a unit test, you're doing everything sequentially: you would call `process` yourself and, I assume, you'd also issue commands directly and not via the command server (so not in a separate thread).

MichaelWedel commented on design/plankton/CommunicationLayerDesign.md:
````diff
@@ -36,13 +36,161 @@ The IBEX team at ISIS, for example, have expressed interest in using Plankton. N
+The `Adapters` in this design are per-protocol & per-device. So they can encapsulate all the logic necessary to drive the specific device that they wrap. 
+
+```python
+class StreamAdapterLinkum95(StreamAdapter):
````
As I said yesterday, I think we can still design base classes for these adapters to mitigate one of the drawbacks you listed below. For example in the `Stream` case I suppose the server part with host and port, collecting stuff that has arrived and sending replies is always the same. The differences are then in how the concrete adapter handles the received input and how it assembles the reply.

OwenArnold replied:
@MichaelWedel Ok. I didn't have much insight into how the base classes could be designed but it would seem sensible to avoid excessive duplication where we need utilities for a particular protocol. I think the base class has two purposes.
1. To avoid code duplication by increasing sharing, which improves maintainability and should make adapter implementations easier.
2. For introspection so that for a given `Device` we can determine what `Adapter(s)` have been made available at ArgParser time.

MichaelWedel replied:
It depends a bit, I think for the EPICS case the base class can be quite "powerful" to make it easy to expose device properties as PVs. For Stream that might be different and the abstraction is "further up", but still it should be possible to eliminate some boiler plate.

Regarding the `ArgParser`-thing, probably the adapters have their own `ArgParser`, like they do now.

Commit:
Added section about package structure

OwenArnold commented on design/plankton/CommunicationLayerDesign.md:
```diff
@@ -190,7 +190,21 @@ I believe this design will satisfy all the listed requirements. But I would be h
+
+* `core`: Contains the state machine and base classes for different adapters, as well as some behavior that is useful for devices (e.g. "approach a value" stuff discussed in [#81](https://github.com/DMSC-Instrument-Data/plankton/issues/81)). These could be in sub-modules/packages `statemachine`, `behavior`, `adapters`, `utilities`.
+* `devices`: These would contain one sub-module per device, at the moment `chopper` and `linkam_t95`, which in turn contain the device implementation as well as the adapter(s) associated with that device. The setups should also be in a sub-module of the device.
+* `devices_contrib` (maybe there is a better name?): This is where external developers would add their devices to begin with. As @owenarnold wrote above, later refactoring done by us (or experienced developers) would move the device into the `devices` package. There is no strict need to have an entire directory per device, it could be a single Python file to begin with.
```
So the structure that you're proposing here gives natural placeholders for device contributions. This is good in that it stops the users having to be modern git developers (know about github PRs etc). Do you have in mind how they might eventually then push those changes to us (We should document this process for our users)? Given that they don't clone a git repository.

MichaelWedel replied:
I'd say once someone is involved enough, they could very well clone the git repository and put their file in the right place. We should definitely put together a "how to contribute" document, but with explanations our "target audience" should be able to do it.

If not, you are right, there should be another channel for contributions. But for the time being it might be enough to be reachable via e-mail?

OwenArnold replied:
Yes. I'd say how to contribute `contributions.md` would be a good addition for git users. PyPI would be updated for major releases to provide newer versions.

MichaelWedel replied:
Great. What is our release schedule? :) At the moment we only really have a bleeding edge version. Maybe we could call what we're producing now (as in after this redesign) version 1.0 (I think it could be a 1.0-thing) and work out proper scopes for subsequent releases afterwards.

OwenArnold committed:
Add low-level separation to my proposal

MichaelWedel commented on design/plankton/CommunicationLayerDesign.md:
````diff
@@ -36,13 +36,161 @@ The IBEX team at ISIS, for example, have expressed interest in using Plankton. N
+        
+```
+
+After discussion with Michael Wedel, I am now strongly leaning towards providing a simple single protocol for the `Environment Control`. While `Adapter` flexibility and variation is vital in order to support different device types. the `Environment` is internal to plankton and we should publish and control a single mechanism for this. Michael Wedel has suggested JSON over zeromq.
````
Today I experimented a bit further with this as a basis, it looks like this can be simplified a bit more. I still have some threading (or rather non-threading) questions, though :)

On 02 Sep 2016, at 16:35, Michael Hart <notifications@github.com<mailto:notifications@github.com>> wrote:

In design/plankton/CommunicationLayerDesign.mdhttps://github.com/DMSC-Instrument-Data/documents/pull/6#discussion_r77355750:

> +
> +```python
> -   class StreamEnvironmentAdapter(object):
>   +
> -    def set_adaptee(self, environment):
> -        self._environment = environment
>   +
>   +
> -    def process(self):
> -        asyncore.loop(0.1, count=1)
> -        # ...
> -        self._environment.set_user_delta(user_value)
>   +
>   +```
>   +
>   +After discussion with Michael Wedel, I am now strongly leaning towards providing a simple single protocol for the`Environment Control`. While`Adapter`flexibility and variation is vital in order to support different device types. the`Environment` is internal to plankton and we should publish and control a single mechanism for this. Michael Wedel has suggested JSON over zeromq.

@MichaelWedelhttps://github.com/MichaelWedel

I took a look at the solution in live-data-prototype, and I really like the idea. I think you're right, this is a better solution than a telnet interface. If we ever really need a telnet interface for something, we could build it on top of it.

In addition to the things Owen mentioned, I'd consider how this might be used from a Python script (like a unit test that wants to control the simulation). It would be good to be able to just import a module and use it as a client directly. The control.py (or whatever we call the command-line interface) could then use that as well / wrap around it.

That's already almost what's there, it's just that:
-   It's a bit awkward that you need two imports / instances (parameter_control_client + general_parameter_controller)
-   Some of the controller's methods print to stdout rather than returning information

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHubhttps://github.com/DMSC-Instrument-Data/documents/pull/6/files/7c43de8f51eff1397ecdcfce70f88350c82696f1#r77355750, or mute the threadhttps://github.com/notifications/unsubscribe-auth/AGETjCwarXIjnKnwnRZNvVBihNZEHugRks5qmDOEgaJpZM4Jv1lV.

MikeHart85 commented on design/plankton/CommunicationLayerDesign.md:
```diff
@@ -36,13 +36,216 @@ The IBEX team at ISIS, for example, have expressed interest in using Plankton. N
+            # Everything still driven from a central processing loop.
+            self._environment_adapter.process()
+            
+            self._adapter.process(delta)
```
Devices may have multiple Adapters active at the same time. Should be just a matter of having a `self._adapters = [...]` and looping over it to call `process` on each one here. But initialization will need to change as well.

Also, as part of "Environment Control", we may want to shut down and start up adapters dynamically at runtime (to simulate a device being physically disconnected and reconnected for example).

Between that and probably needing to list available adapters for command-line / argparsers purposes, we should think about how we might query what adapters a given device supports. And also get an instance of an adapter for device X and protocol Y, without needing the exact name of the class.

OwenArnold replied:
_we should think about how we might query what adapters a given device supports_
- Yes, the query should be easy enough. The adapter wraps the device and know about what protocol it supports 
  so all the information is there.

_Will we ever have / should we support more than one Adapter for the same protocol and the same device_
- As far as I understand the requirements - probably we won't need to. But I don't see a valid reason to restrict it either. The possibility of having more than one adapter for the same device seems to logically drop out of the separation of protocols from devices.

MikeHart85 replied:
I'm thinking about before we create an instance of an adapter, to know what adapters are available and which classes implement them / where to find them. For example, if the user wants to list all adapters for the T95 using something like:

`$ ./plankton.py -d linkam_t95 --list-adapters`

Or the user the starts a device with some adapter:

`$ ./plankton.py -d chopper -p stream`

How do we know whether or not `chopper` has a `stream` adapter, and where to find it?

OwenArnold replied:
`$ ./plankton.py -d linkam_t95 --list-adapters` @MikeHart85 Yes. That's what I have in mind too.  I fear that I may be straying too far into specifics here, but with my design it's pretty straight-forward. We have a known package structure already for plankton. Adapters are bespoke to devices in my proposal, so they should live in some standard way near their device. Adapter discovery is therefore simply a matter of inspecting directories. An alternative would be for the Adapter to provide a method something like (`adaptsDevice(self, device)` returning bool), but I think that's less pythonic. To know what sort of Adapter a given Adapter type is we simply need to run `isinstance` i.e. `isinstance(adapter, StreamAdapter)` see `Adapters` subheading.

MichaelWedel replied:
We already have these `import_adapter`-methods and so on, they probably need to be re-written slightly, but basically they are the same. The fact that each device will come with its own set of adapters will make this easier, because we don't need to check "does this combination of device and adapter exist" (we're not doing that at the moment either, everything just fails spectacularly) but only "what adapters does this device come with". One thing we need to think about is naming though. I wouldn't want users to type `-d chopper -p ChopperAdapterEpics`, so we need to tackle this somehow. One option (@owenarnold, you're going to love this) would be to have an `AdapterFactory` where we could subscribe adapters per device. It would look something like this `factory.subscribe(DeviceClass, AdapterClass, 'protocolname')` and then you could create the proper adapter via `factory.createAdapter(device_object, 'protocolname')`.

OwenArnold replied:
@MichaelWedel - `AdapterFactory`  A little part of me died when you wrote that. :)

MichaelWedel replied:
I know Mantid kind of went too far on some of this, but the factory pattern seems like a reasonable thing to implement here, since "string based creation of stuff" makes it easier to access all this. I'd even go as far as proposing to use a singleton for this, (i.e. have one factory object somewhere that can be imported `from plankton import AdapterFactory` or something like that.

OwenArnold replied:
Yes, it does seem to fit better here. And yes, Mantid went far too far without properly addressing scope and creational issues. That's why @SimonHeybrock is having so much fun at the moment.

MikeHart85 replied:
I'm conflicted about the Factory approach (and doubly so about using a singleton... anywhere... for anything... ever).

It was my first thought as well when asking this question. "Want to generate instances -> use factory" must be hardwired somewhere in my brain.

But where would we put the `factory.subscribe` calls given that this is Python? 

Can't put it in the adapters' `__init__`, cause we're not instantiating them unless we need them. Can't put it elsewhere in the same .py file, because we won't necessarily be importing it at all. So now we have to put them in some central place that always gets executed... but then we're effectively spelling out a mapping of device + protocol names to classes... why not make a `dict` and not bother with a factory? 

Maybe it's best to avoid that route entirely and just say all adapters for a device should go for a directory/filename approach. So put all the adapters into a subfolder under the device with the protocol as the name:

```
plankton
- devices
 |- chopper
   |- adapters
   | |- epics.py
   | |- stream.py
   | |- modbus.py
   |- device.py
```

But that would mean only one adapter per protocol. Alternatively, we could allow arbitrary names for the adapter .py files, but then we would have to load all of them and somehow figure which protocols each of the classes we find supports (perhaps just look at what they inherit from?).

But if we allow multiple adapters per protocol, then what does `$ plankton.py -d chopper -p stream` mean?

Anyway, this is probably getting too far into implementation details for this document... just some things to think about.

MichaelWedel replied:
Regarding the factory, I'd say it wouldn't store Adapter instances, it would store Adapter-classes and the factory would create an instance of the requested class. So subscription would happen in the `epics.py` etc. files (right where the adapter-class is specified).

```
from somewhere import AdapterFactory
from ..device import Chopper

class ChopperEpicsAdapter(EpicsAdapter):
    # define wonderful adapter

AdapterFactory.subscribe(Chopper, ChopperEpicsAdapter, 'epics')
```

As far as multiple adapters go I'd say we could allow something like `-d chopper -p epics stream modbus` (via `add_argument('-d', nargs='*')`), the only thing I'm not sure about would be passing on parameters to all these adapters...

MikeHart85 replied:
But that...

```
AdapterFactory.subscribe(Chopper, ChopperEpicsAdapter, 'epics')
```

... would only be executed when you import that specific `epics.py`. And we want those imports to happen dynamically, right? If not, where would the `import device.chopper.adapters.epics` live? And why bother subscribing if we know that's where we'll find the epics adapter for the chopper?

MikeHart85 commented on design/plankton/CommunicationLayerDesign.md:
```diff
@@ -36,13 +36,216 @@ The IBEX team at ISIS, for example, have expressed interest in using Plankton. N
+            # Everything still driven from a central processing loop.
+            self._environment_adapter.process()
+            
+            self._adapter.process(delta)
```
Related thought: Will we ever have / should we support more than one Adapter for the same protocol and the same device?

MikeHart85 commented on design/plankton/CommunicationLayerDesign.md:
```diff
@@ -36,13 +36,216 @@ The IBEX team at ISIS, for example, have expressed interest in using Plankton. N
+            return (datetime.now() - start).total_seconds()
+            
+    def set_user_delta(self, delta):
+        self._user_delta = delta
```
Talked about this with Owen last week, but forgot to leave a comment here: The delta t would be controlled by a multiplier (which defaults to 1) rather than a fixed override as shown here. Not sure if we want to update the document or just consider that an implementation detail.

Commit:
Added code example section

MichaelWedel committed:
Included short summary of decision

OwenArnold committed:
complete approval table

OwenArnold committed:
Change section to remove alternative

The section used an alternative design as an illustration. This is not the accepted design as of the 5th september, and it causes confused and added some confusing new terminology. I've stripped those parts out.
